src/main/java/org/commcare/formplayer/services/RestoreFactory.java

```diff
@@ -611,37 +616,45 @@ private HttpHeaders getHmacHeaders(String requestPath) {
         builder.append("/");
         HttpHeaders headers = getHmacHeaders(builder.toString());
         String fullUrl = host + builder.toString();
-        return new Pair<> (fullUrl, headers);
+        return new Pair<> (UriComponentsBuilder.fromUriString(fullUrl).build(true).toUri(), headers);
     }
-    public Pair<String, HttpHeaders> getUserRestoreUrlAndHeaders() {
+    public Pair<URI, HttpHeaders> getUserRestoreUrlAndHeaders() {
         return getUserRestoreUrlAndHeaders(false);
     }
 
-    public Pair<String, HttpHeaders> getUserRestoreUrlAndHeaders(boolean skipFixtures) {
-        StringBuilder builder = new StringBuilder();
-        builder.append("/a/");
-        builder.append(domain);
-        builder.append("/phone/restore/?version=2.0");
+    public Pair<URI, HttpHeaders> getUserRestoreUrlAndHeaders(boolean skipFixtures) {
+        // URI
+        String restoreUrl = "/a/" + domain + "/phone/restore/?version=2.0";
+        String uri = host + restoreUrl;
+        UriComponentsBuilder builder = UriComponentsBuilder.fromUriString(uri);
         String syncToken = getSyncToken();
         if (syncToken != null && !"".equals(syncToken)) {
-            builder.append("&since=").append(syncToken);
+            builder.queryParam("since", syncToken);
         }
-        builder.append("&device_id=").append(getSyncDeviceId());
-
+        builder.queryParam("device_id", getSyncDeviceId());
         if (useLiveQuery) {
-            builder.append("&case_sync=livequery");
+            builder.queryParam("case_sync", "livequery");
         }
-
         if( asUsername != null) {
-            builder.append("&as=").append(asUsername);
+            String unEncodedAsUsername = asUsername;
```

Comment:
@snopoke I see you made some changes to this logic just now, can you check my modified changes.

Reply:
LGTM